### PR TITLE
add controller to maintain default SA permissions

### PR DESCRIPTION
**Pull request by deads2k**
We "normally" maintain permissions for image pusher, image puller, and deployer using new-project templates and post-start hooks, but as we add automated components that are not openshift aware and still want features like deploymentconfigs to work, the added maintenance burden seems unnecessary.

This pull introduces a controller which does our bindings for SAs to pull images in their project, allows the builder SA to push images in the project, and allows the deployer SA to create pods and RCs.  It is still possible to opt out if you really want to break openshift features in your project by removing the subjects from the binding (I only match on name).

No tests yet.  Let's agree in concept first, then I'll add them before we merge.

@openshift/sig-master

### CODE REVIEW

**Comment by deads2k** [comment]
tests added

**Comment by soltysh** [comment]
@deads2k is this going to save us from bugs like https://github.com/openshift/api/issues/35 ?

**Comment by deads2k** [comment]
> @deads2k is this going to save us from bugs like openshift/api#35 ?

heh. I had no idea.

**Comment by simo5** [comment]
@enj PTAL

**Comment by deads2k** [comment]
Green!

/assign @enj

File: pkg/project/registry/projectrequest/delegated/sample_template.go [hunk 1]
```diff
@@ -43,6 +43,7 @@ func DefaultTemplate() *templateapi.Template {
 		panic(err)
 	}
 
+	// TODO this should be removed in 3.11.  We need to keep it for new server, old controller cases in 3.10.
```
Comment (enj) [inline pkg/project/registry/projectrequest/delegated/sample_template.go]: Issue please.

File: pkg/authorization/controller/defaultrolebindings/defaultrolebindings.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+func NewDefaultRoleBindingsController(roleBindingInformer rbacinformers.RoleBindingInformer, namespaceInformer coreinformers.NamespaceInformer, roleBindingClient rbacclient.RoleBindingsGetter) *DefaultRoleBindingController {
+	c := &DefaultRoleBindingController{
+		roleBindingClient: roleBindingClient,
+
+		roleBindingLister: roleBindingInformer.Lister(),
+		roleBindingSynced: roleBindingInformer.Informer().HasSynced,
+		namespaceLister:   namespaceInformer.Lister(),
+		namespaceSynced:   namespaceInformer.Informer().HasSynced,
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ClusterRoleAggregator"),
```
Comment (enj) [inline pkg/authorization/controller/defaultrolebindings/defaultrolebindings.go]: Wrong name

File: pkg/authorization/controller/defaultrolebindings/defaultrolebindings.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+	for i := range desiredRoleBindings {
+		desiredRoleBinding := desiredRoleBindings[i]
+		found := false
+		for _, existingRoleBinding := range roleBindings {
+			if existingRoleBinding.Name == desiredRoleBinding.Name {
+				found = true
+				break
+			}
+		}
+		if found {
```
Comment (enj) [inline pkg/authorization/controller/defaultrolebindings/defaultrolebindings.go]: > It is still possible to opt out if you really want to break openshift features in your project by removing the subjects from the binding (I only match on name).

This is an interesting choice.  Why not the standard pattern of auto reconcile unless you have the protect annotation?

File: pkg/authorization/controller/defaultrolebindings/defaultrolebindings_test.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	kubeclientfake "k8s.io/client-go/kubernetes/fake"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	rbaclisters "k8s.io/client-go/listers/rbac/v1"
+	clienttesting "k8s.io/client-go/testing"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kubernetes/pkg/controller"
+)
+
+func TestSync(t *testing.T) {
```
Comment (enj) [inline pkg/authorization/controller/defaultrolebindings/defaultrolebindings_test.go]: No tests for fixup of deletions?

Comment (deads2k) [reply pkg/authorization/controller/defaultrolebindings/defaultrolebindings_test.go]: > No tests for fixup of deletions?

The way I wrote the controller, that case looks identical to any other, "I don't have given rolebinding"

File: pkg/authorization/controller/defaultrolebindings/defaultrolebindings.go (hunk 3 above)
Comment (deads2k) [reply pkg/authorization/controller/defaultrolebindings/defaultrolebindings.go]: > This is an interesting choice. Why not the standard pattern of auto reconcile unless you have the protect annotation?

It's just unnecessary. Presence or absence is enough.  Figure that you'd have to create it anyway to annotate it.

File: pkg/authorization/controller/defaultrolebindings/defaultrolebindings.go (hunk 2 above)
Comment (deads2k) [reply pkg/authorization/controller/defaultrolebindings/defaultrolebindings.go]: fixed

File: pkg/project/registry/projectrequest/delegated/sample_template.go (hunk 1 above)
Comment (deads2k) [reply pkg/project/registry/projectrequest/delegated/sample_template.go]: > Issue please.

https://github.com/openshift/origin/issues/18923

**Comment by deads2k** [comment]
comments addressed.

**Pull request merged**